### ISSUE TITLE
Long-term Geographic Features planning

### ISSUE THREAD

**Issue by jywarren** [publiclab/plots2#1416]
This is not a regular issue, but is a long-term tracking issue for all Geographic features. It's related to work on this repository and to https://github.com/publiclab/leaflet-blurred-location/issues/1

****

# Release v0.1 - bare bones minimal useful release

1. [x] inline maps: [map:content:lat:lon] or [map:nodes:lat:lon] (just nodes)
2. [x] load half a degree box around map center - started with snipping the last digit (i.e. 71.0 for "lat:71.05") to show the next zoom level up from specified center
3. [x] Leaflet map
4. [x] maps of multiple things, like [map:all:lat:lon] (relates to 7)
5. [x] maps PER TAG: [map:tag:______:lat:lon] ??
6. [ ] default display on each tag page, like https://publiclab.org/tag/stormwater or https://publiclab.org/tag/air-quality - under a tab?
7. [x] zoom (ability to specify custom initial zoom level)

****

# Release v0.2 - "people and privacy"

## Location entry

(mostly via https://github.com/publiclab/leaflet-blurred-location/issues/1)

1. [ ] standardize for use in nodes or profiles via separate module for BlurredLocation
1. [ ] creates tags "lat:____", "lon:_____", "blurred:true", "place:buffalo-ny"
1. [ ] use tag description field for full placename? So "place:buffalo-ny" would have description; "Buffalo, New York, ZIPCODE, USA" ?
1. [ ] manual entry "place:buffalo-ny" auto-generates "lat:xxx" + "lon:xxx" - at what zoom level? "City"? Check google geocoder precision field?
1. [ ] parameterize location as you type, so you see "buffalo-ny" appear as you type "Buffalo, NY", and this is submitted as `place:buffalo-ny` with description "Buffalo, New York, ZIPCODE, USA"
1. [ ] display tag desc on tag page
1. [ ] display tag desc on tag hover (made into FTO at #1406)

## Map displays

1. [ ] Upgrade: bounding box request on drag or zoom (requires Map API section)
2. [ ] Search box with geolocation and map refresh
3. [ ] Advanced short codes for inline maps:
    a. [ ] (v0.2) [map:people:lat:lon(:zoom)] (just people: requires privacy blurring)
    b. [ ] (v0.2 or v0.3) variations like [map:questions:_____],  [map:activities:_____]
    c. [ ] (optional) [map:content:Buffalo, NY] (can this work?) or [map:content]Buffalo, NY[/map]
4. [ ] Map display on additional standard pages: /people or /questions
5. [ ] display of blurred locations: 
    a. [ ] when does it start being a point? when you zoom out far enough?
    b. [ ] no longer display an area for 41.03,71.04 if you zoom past xx.xx resolution?
    c. [ ] start out by not displaying blurred locations? Blurred in release 2?
6. [ ] toggle-able map Key: by type (for "all" map type)

## Map API

1. [ ] way to request located nodes or profiles in GeoJSON
2. [ ] JSON content at URL: /tag/tagname.json?geo=true&lat1=__&lat2=__&lon1=__&lon2=__ (bounding box)
3. [ ] eventually allow for natural language queries like /geo/q=Buffalo, NY ?
4. [ ] relying on calls of this type will also make pulling in JusticeMap content easier

## Geo Subscriptions

1. [ ] Relate to GSoC summer email notifications
2. [ ] area-based tag-based subscriptions? Perhaps via wildcard subscriptions?
3. [ ] "near me" or "near X"

**Comment by mridulnagpal** [comment]
@jywarren Shall we start with the integration?

**Comment by jywarren** [comment]
Sure, how's the display library going?

For interface, I'd like to suggest you try to get the basic input to
display in a partial template, as you originally did, but this time using
the new library!

**Comment by mridulnagpal** [comment]
The display library is at a prototype stage now, it builds fine and has some  basic methods.

**Comment by mridulnagpal** [comment]
We need to add the leaflet-blurred-location library to npm first, or we just link it from github itself?

**Comment by jywarren** [comment]
Can you confirm the Leaflet Blurred Location package.json is up to date and
the version is correct? If so i can publish it.

**Comment by mridulnagpal** [comment]
The version is 1.1.0 and the name is leaflet-blurred-location
Should I change anything?

**Comment by jywarren** [comment]
I trust your call -- should i grab the latest master and publish it? :-)

**Comment by jywarren** [comment]
added integration steps from #1655 !!

**Comment by mridulnagpal** [comment]
@jywarren Shall we get on with this now? What's our priority for now?

**Comment by mridulnagpal** [comment]
@jywarren Let's finish off with this milestone for starters, or you have something else in mind?

**Comment by jywarren** [comment]
Cleaned up and updated some tasks!!!

**Comment by mridulnagpal** [comment]
So basically, we enter place: buffalo-ny, it then converts to  "Buffalo, New York, ZIPCODE, USA" and then does a search. Right?

**Comment by jywarren** [comment]
so this would be in the tag input on a PublicLab.org page --

so we'd detect a `place:____` tag here:
https://github.com/publiclab/plots2/blob/0647cb0a7549821f7442213b9571dddfaebedcbb/app/controllers/tag_controller.rb#L278

Then we'd replace '-' with ' '

Then we'd use the already-installed Geocoder gem to try to find where it
is: https://github.com/alexreisner/geocoder

Then we'd add two additional tags: `lat:____` and `lon:____`.

What do you think? Should we instead try to catch this in JavaScript and
have a popup that says "Looks like a location -- is this right?" and shows
lat/lon? That'd give people a chance to agree/disagree.

Or we could even just let people add the tag, but then show an alert above
the tag input area that says "That looked like a location -- _click here_
to add coordinates."

Which do you think is the best experience and the most compact code?

**Comment by mridulnagpal** [comment]
The option to add location co-ordinates sounds better, as all the location won't be converted to lat lon as well, thus we could give a pop up as you said.

**Comment by mridulnagpal** [comment]
@jywarren Also if we use javascript we can sync the co-ordinates with the leaflet-blurred-location library as well, using the same google API we used in leaflet-blurred-location

**Comment by mridulnagpal** [comment]
@jywarren Any updates?

**Comment by jywarren** [comment]
Ok so if we do it in JavaScript then we won't need the place tag dash removal in ruby code - we can do that in JavaScript! We will need to detect place tags being added in JavaScript though. I believe the file is `app/assets/javascripts/tagging.js`

**Comment by mridulnagpal** [comment]
There is a method addTag with tagname and selector as input, we can directly check it there right, if its a place? Also when I do console log in the addTag method it doesn't display on the console when I publish the doc. How do I debug ?

**Comment by jywarren** [comment]
Yes, that makes sense -- add a filter inside `addTag()`. I believe `addTag()` uses the HTML form on the page, rather than making an independent AJAX call. We might try to switch to doing an AJAX call, but I guess it really seems to work OK right now, so why bother?

But consider -- `addTag()` relies on this code being present in HTML: https://github.com/publiclab/mapknitter/blob/master/app/views/tags/_index.html.erb

In any case, on the `tag_controller` side, it's OK to submit several tags at once, separated by commas, like `one,two,three:etc` -- so you could:

1. [ ] intercepts `addTag()` and if it has `place:___`, then 
2. [ ] replace `-` with ` ` (a space) and try to geocode
3. [ ] if successful, pop up a modal saying "we've detected a location and tried to geocode it; does this look right?" showing a longer text description of the place and buttons of `Yes, add coordinates` and `No, just use the original tag` options
4. [ ] if they click `Yes`, add `,lat:_____,lon:______` to the submitted tag string, else just send the original `place:_____` tag alone

How does this sound?

**Comment by mridulnagpal** [comment]
#2734 Please have a look

**Comment by mridulnagpal** [comment]
@jywarren Whats next?

**Comment by jywarren** [comment]
Hi Mridul there are a few open issues in the blurred image repo, want to
try those and also some CodeClimate based code structure work?

Thanks!!!

On Jun 5, 2018 4:15 PM, "Mridul Nagpal" <notifications@github.com> wrote:

@jywarren <https://github.com/jywarren> Whats next?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/publiclab/plots2/issues/1416#issuecomment-394877723>,
or mute the thread

**Comment by daemon1024** [comment]
Is this the meta issue for https://publiclab.org/wiki/gsoc-ideas#Geographic+features+refinements ? I wanted to discuss the proposed features, would this be the correct thread to get insights on it?

**Comment by jywarren** [comment]
Hi @daemon1024 yes it is, but it's SOOOOO old! I wonder if we should start anew? Many above ideas are done or discarded. What do you think?

**Comment by daemon1024** [comment]
@jywarren Sure. Let's start up a fresh planning issue.
